test/unit/modules/blocks/verify.js

```diff
@@ -0,0 +1,635 @@
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Failed to verify block signature');
+			done();
+		});
+
+		it('verify block signature should fail (invalid generatorPublicKey: no hex)', function (done) {
+			invalidBlock.blockSignature = validBlock1.blockSignature;
+			invalidBlock.generatorPublicKey = 'invalidblocksignature';
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('TypeError: Invalid hex string');
+			done();
+		});		
+
+		it('verify block signature should fail (invalid generatorPublicKey: hex)', function (done) {
+			invalidBlock.generatorPublicKey = '948b8b509579306694c00db2206ddb1517bfeca2b0dc833ec1c0f81e9644871b';
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Failed to verify block signature');
+			done();
+		});
+
+		it('calculate expected rewards should fail (invalid reward)', function (done) {
+			invalidBlock.reward = 555;
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal(['Invalid block reward:', invalidBlock.reward, 'expected:', validBlock1.reward].join(' '));
+			done();
+		});
+		
+		it('total fee should fail (invalid total fee)', function (done) {
+			invalidBlock.totalFee = 555;
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Invalid total fee');
+			done();
+		});
+		
+		it('payloadHash should fail (invalid payload hash)', function (done) {
+			invalidBlock.payloadHash = 'invalidpayloadhash';
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Invalid payload hash');
+			done();
+		});
+
+		it('transactions check should fail (duplicate transaction)', function (done) {
+			invalidBlock.transactions.push(invalidBlock.transactions[0]);
+			invalidBlock.numberOfTransactions = 2;
+			
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Encountered duplicate transaction: ' + invalidBlock.transactions[1].id);
+			done();
+		});
+
+		it('transactions check should fail (getBytes(): Unknown transaction type)', function (done) {
+			invalidBlock.transactions[0].type = 555;
+
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Unknown transaction type ' + invalidBlock.transactions[0].type);
+			done();
+		});
+
+		it('transactions check should fail (length is too high)', function (done) {
+			invalidBlock.transactions[0].type = validBlock1.transactions[0].type;
+			invalidBlock.transactions = new Array(26);
+			invalidBlock.numberOfTransactions = invalidBlock.transactions.length;
+			
+			var check = blocksVerify.verifyBlock(invalidBlock);
+			expect(check).to.equal('Transactions length is too high');
+			done();
+		});
+
+		it('transactions check should fail (number of transactions)', function (done) {
+			invalidBlock.transactions = validBlock1.transactions;
```

Comment:
We are mutating `invalidBlock` all the way, not clean what is wrong with that block at current step. Maybe use `validBlock` and change at each step only tested properties?